router/broker.go

```diff
@@ -898,3 +1011,138 @@ func (b *broker) subCountSubscribers(msg *wamp.Invocation) wamp.Message {
 		Arguments: wamp.List{count},
 	}
 }
+
+// eventHistoryLast retrieves events history for subscription applying provided filters
+// TODO: Need to filter by authid and/or authrole of caller if stored events have any of them
+// But that's not possible in current arch as meta peer doesn't know anything about caller, just invocation message
+func (b *broker) subEventHistory(msg *wamp.Invocation) wamp.Message {
+	var events wamp.List // []*storedEvent
+	var isLimitReached bool
+	var fromDate time.Time
+	var afterDate time.Time
+	var beforeDate time.Time
+	var uptoDate time.Time
+	var dateStr string
+	var topicStr string
+	var topicUri wamp.URI
+	var err error
+
+	subId, ok := wamp.AsID(msg.Arguments[0])
+
+	if !ok {
+		return &wamp.Error{
+			Type:    msg.MessageType(),
+			Request: msg.Request,
+			Details: wamp.Dict{},
+			Error:   wamp.ErrInvalidArgument,
+		}
+	}
+
+	limit, ok := msg.ArgumentsKw["limit"].(int)
+	if ok && limit < 1 {
+		return &wamp.Error{
+			Type:    msg.MessageType(),
+			Request: msg.Request,
+			Details: wamp.Dict{},
+			Error:   wamp.ErrInvalidArgument,
+		}
+	}
+
+	dateStr, ok = msg.ArgumentsKw["from"].(string)
+	if ok {
+		fromDate, err = time.Parse(time.RFC3339, dateStr)
+		if err != nil {
+			return &wamp.Error{
+				Type:    msg.MessageType(),
+				Request: msg.Request,
+				Details: wamp.Dict{},
+				Error:   wamp.ErrInvalidArgument,
+			}
+		}
+	}
+
+	dateStr, ok = msg.ArgumentsKw["after"].(string)
+	if ok {
+		afterDate, err = time.Parse(time.RFC3339, dateStr)
+		if err != nil {
+			return &wamp.Error{
+				Type:    msg.MessageType(),
+				Request: msg.Request,
+				Details: wamp.Dict{},
+				Error:   wamp.ErrInvalidArgument,
+			}
+		}
+	}
+
+	dateStr, ok = msg.ArgumentsKw["before"].(string)
+	if ok {
+		beforeDate, err = time.Parse(time.RFC3339, dateStr)
+		if err != nil {
+			return &wamp.Error{
+				Type:    msg.MessageType(),
+				Request: msg.Request,
+				Details: wamp.Dict{},
+				Error:   wamp.ErrInvalidArgument,
+			}
+		}
+	}
+
+	dateStr, ok = msg.ArgumentsKw["up_to"].(string)
+	if ok {
+		uptoDate, err = time.Parse(time.RFC3339, dateStr)
+		if err != nil {
+			return &wamp.Error{
+				Type:    msg.MessageType(),
+				Request: msg.Request,
+				Details: wamp.Dict{},
+				Error:   wamp.ErrInvalidArgument,
+			}
+		}
+	}
+
+	topicStr, ok = msg.ArgumentsKw["topic"].(string)
+	if ok {
+		topicUri = wamp.URI(topicStr)
+	}
+
+	ch := make(chan struct{})
+	b.actionChan <- func() {
+		var filteredEvents []storedEvent
+
+		if subscription, ok := b.subscriptions[subId]; ok {
+			if storeItem, ok := b.eventHistoryStore[subscription]; ok {
+				isLimitReached = storeItem.isLimitReached
+
+				for _, event := range storeItem.events {
+					eventTopic, ok := event.event.Details["topic"]
+
+					if (fromDate.IsZero() || !event.event.timestamp.Before(fromDate)) &&
```

Comment:
Since you iterate, I would switch this to the following style for readability:
``` go
if !fromData.IsZero() && event.event.timestamp.Before(fromDate) {
   continue
}
```

This way you can avoid the very long if sequence, that is somewhat hard to mentally parse.